documentation/SOFTWARE_DESIGN.md

```diff
@@ -65,7 +65,9 @@ Sokoban constructor(const vec<vec<string>> levels)
 // class properties
 const vec<vec<string>> levels, read-only
 uint current_level // index into levels
-uint moves
+vector<Direction> moves // history of moves made by the player
```

Comment:
When I try to code this, I get an error when I use the same name for two different data types.

Reply:
Good call, `move_history` seems fine instead to differentiate from the getter.

Yes, a string wouldn't be the right data type for move history even though I guess they're mutable in C++. One thing I noticed in your draft implementation is that there's a lot of converting between square characters and enums, and the same thing for directions. If we find we have to convert these enums back and forth between strings/chars all the time, it sort of defeats the purpose of these abstractions and just becomes a liability. If the enums turn out to be too much work to manage, I'd vote to just go all-chars/strings or whatever and just agree not to use any invalid strings with the client. This happens all the time, so I left it sort of ambiguous in the design -- it should evaluate to "use whatever makes sense and is the least amount of pain to deal with".

Back to move history, I realize now that I actually pushed the entire game state on each move back in my original Sokoban https://github.com/ggorlen/sokoban/blob/master/js/sokoban.js#L87. That's less efficient, but given that undos don't need to be hyper-performant (until we get to writing AI, which we probably won't do because it's very difficult), it might be the best way to go for us. The problem with only keeping track of directions is that when doing an undo, we have to reverse the logic for pushing boxes, which is not necessarily trivial and involves basically re-writing the hardest function in the game to work in two directions. I recall that being a lot of extra work, which is why I went for the lazy approach.

So I'm thinking we can just have a vetor of whole board states and pop it, something like `std::vector<Sokoban>` basically... thoughts based on your experiments so far?

Reply:
I didn't like the idea of saving the entire game state. It may consume a lot of space.  If time is an issue then going with the lazy way is our best approach. But I am willing to put some work and time to make improvements on the undo() function which I am experimenting with right now. I have two ideas of how to do this:

a. I thought of just saving a history of a two-cell snapshot of every move taken so far. It could be a struct that contains Points, Items. Point 1 is the coordinate of item 1 and Point 2 is the coordinate of item 2. Then we can just pop this snapshot and print it back to the board. We might need another function display-board() to print every piece of the board every time a state changes. But it might require more time to implement. I'll have to think about this more.

b.I am leaning towards creating two functions: push() and pull() to handle the logic of how the player is interacting with the box. Then use these functions to incorporate to the move() function. Maybe we need to add a parameter that can activate the pull function when undo is being called. 

I was reading about the AI capabilities that solve the entire board. And it has an NP-hard category. People are still trying to make efficient ways of solving this. I definitely will not be going in that route (at least for now). But I still want to explore the idea of applying some kind of BFS or DFS algorithm for finding the optimum path to a solution. So if a solution is known, we can compare that to the current player's state and suggest optimal sequences based on the moves the player has taken so far.

Reply:
> I didn't like the idea of saving the entire game state. It may consume a lot of space.

It's generally best to implement things the inefficient, easy way, then go for the harder but more efficient way later only after a real performance problem that affects users arises. Memory is cheap, so we can use it without worry. [This article](https://ericlippert.com/2012/12/17/performance-rant/) might help.

Also, it shouldn't be too hard to refactor it if it's done correctly--saving and restoring state should be a total of two functions that operate on one data structure, totally isolated from the rest of the logic. So my suggestion is to do it the easy/dumb way, get it working, then refactor to the more difficult and elegant approach later if there's time.

Finally, we have limited time, so the tradeoff is between optimizing memory consumption versus implementing cooler features and improving the game in other ways (as Eric's article states) that visibly benefit users, I'd go with the latter.

> But I still want to explore the idea of applying some kind of BFS or DFS algorithm for finding the optimum path to a solution

You're welcome to give it a try, but I suggest we keep this out of the team roadmap. I [tried](https://github.com/ggorlen/sokoban/blob/master/js/sokosolver.js) this and pretty much gave up.

It's a very difficult problem -- for a single move, you need a BFS per box just to find the shortest path to each side of each of the boxes, then you have to try pushing each of the boxes in every possible direction it can be pushed in, then run a function to see which pushes have led to impossible-to-solve positions so you can backtrack. So for every move, we have, say, 10 or 20 different child positions. For each one of those, we have 10 or 20 child positions and that's an exponential blowup. We also need to detect cycles and avoid repeating the same state. It's a mess but if you want to try, it'll be educational! I might be interested in collaborating on it in the future though, but I'd rate this as "extremely hard" for the purposes of our project.

Reply:
> Also, it shouldn't be too hard to refactor it if it's done correctly--saving and restoring state should be a total of two functions that operate on one data structure, totally isolated from the rest of the logic. So my suggestion is to do it the easy/dumb way, get it working, then refactor to the more difficult and elegant approach later if there's time.

> Ok, I'm ok with this. This will make a lot of the implementation easier. 

> Finally, we have limited time, so the tradeoff is between optimizing memory consumption versus implementing cooler features and improving the game in other ways (as Eric's article states) that visibly benefit users, I'd go with the latter.
> I agree. I'd rather work with the cooler features. I can't wait.

> > But I still want to explore the idea of applying some kind of BFS or DFS algorithm for finding the optimum path to a solution
> 
> You're welcome to give it a try, but I suggest we keep this out of the team roadmap. I [tried](https://github.com/ggorlen/sokoban/blob/master/js/sokosolver.js) this and pretty much gave up.
> 
> It's a very difficult problem -- for a single move, you need a BFS per box just to find the shortest path to each side of each of the boxes, then you have to try pushing each of the boxes in every possible direction it can be pushed in, then run a function to see which pushes have led to impossible-to-solve positions so you can backtrack. So for every move, we have, say, 10 or 20 different child positions. For each one of those, we have 10 or 20 child positions and that's an exponential blowup. We also need to detect cycles and avoid repeating the same state. It's a mess but if you want to try, it'll be educational! I might be interested in collaborating on it in the future though, but I'd rate this as "extremely hard" for the purposes of our project. Even if we were able to code up the above, then it probably won't solve any of the original Sokoban levels in any sane amount of time, without serious optimization efforts. If memory serves me correct, solvers typically can't use BFS or A* because of memory, and wind up using IDDFS, for starters.
> True, BFS requires a LOT of memory space. I think you are right. This is extremely difficult. One of the features that I am interested in is doing HINT. Basically, what it does is it will undo all the moves that you have done up to the point of the farthest location that you already solved correctly. And then reveal the solution path of solving one box. 

>Are the solutions sequential? Do putting the boxes in their places have to be in order? Is there more than one optimal solution?

Reply:
> Solutions can be in any order, and putting one box on a goal doesn't necessarily mean the box is permanently solved. Frequently, you'll temporarily "solve" a box, then later move it to another goal, or move it back off the goal. Or, putting a box on a goal needs to be deferred until a more optimal position is reached by reshuffling other boxes in arbitrary configurations. Which box to put on which goal matters greatly, so even if this was possible you'd have to try putting every box on every goal -- and then solve the rest of the game to prove that it was actually a productive move.

Ok, this convinces me to abandon the idea of a HINT feature. After watching the video and reading this thread, pursuing this Sokoban solver route is definitely out of scope for this project. 

> At [2:55](https://youtu.be/YX40hbAHx3s?t=175), the presenter talks about how if I ask you for the best move in a chess position and you give me an answer, how do I even check that it's the best move? A similar idea applies to Sokoban hinting as far as I can tell.

Yes, this is a good question to raise.

> If we wanted to go into AI, something like Connect 4, Reversi or Checkers would have been the way to go. These AIs are much more doable, typical Monte Carlo or Minimax approach works, and the coding to manipulate the board is easy. For puzzles like peg solitaire and Klotski, a simplistic brute force search will generally be sufficient to solve a small puzzle, so these are in scope as well.
> Also relevant: https://xkcd.com/1425/

Ok, I'll take a lot at this. This is great info.